### PR TITLE
Poseidon hash

### PR DESCRIPTION
**Pull request by tsunrise**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -16,17 +16,36 @@ exclude = [
...
-dusk-bls12_381 = "0.8"
-dusk-jubjub = "0.10"
-dusk-schnorr = "0.7.1"
-dusk-poseidon = { version = "0.22", features = ["canon"] }
-dusk-bytes = "0.1"
+rand_core = "0.6.0"
+rand = "0.8"
+thiserror = "1.0.30"
+
+ark-plonk = { version = "^0.8", git = "https://github.com/rust-zkp/ark-plonk", default-features = false }
```
Comment (stechu) [inline Cargo.toml]: Is it possible to update to the new repo? https://github.com/ZK-Garage/plonk

Comment (tsunrise) [reply Cargo.toml]: yes, updated

File: src/poseidon/matrix.rs [hunk 2]
```diff
@@ -0,0 +1,675 @@
+//! acknowledgement: adapted from FileCoin Project: https://github.com/filecoin-project/neptune/blob/master/src/matrix.rs
+use std::iter::FromIterator;
+use std::ops::{Index, IndexMut};
+
+use ark_ff::PrimeField;
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct Matrix<T: Clone>(pub Vec<Vec<T>>);
```
Comment (bhgomes) [inline src/poseidon/matrix.rs:8]: I would suggest using a `Vec<T>` in row-major order for better performance.

File: src/poseidon/matrix.rs [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+    pub fn identity(n: usize) -> Matrix<F> {
+        let mut m = Matrix(vec![vec![F::zero(); n]; n]);
+        for i in 0..n {
+            m.0[i][i] = F::one();
+        }
+        m
+    }
+
+    pub fn is_identity(&self) -> bool {
+        if !self.is_square() {
```
Comment (bhgomes) [inline src/poseidon/matrix.rs:94]: Aren't all these matrices square? The transpose method for example seems to assume so. Do we need non-square matrices anywhere?

File: src/poseidon/poseidon_ref.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+use crate::poseidon::mds::MdsMatrices;
+use crate::poseidon::PoseidonError;
+use crate::poseidon::round_constant::generate_constants;
+use crate::poseidon::round_numbers::calc_round_numbers;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct PoseidonConstants<F: PrimeField> {
+    pub mds_matrices: MdsMatrices<F>,
+    pub round_constants: Vec<F>,
+    pub domain_tag: F,
```
Comment (bhgomes) [inline src/poseidon/poseidon_ref.rs]: What is the `domain_tag` used for?

**Review by bhgomes — COMMENTED** [review]
Just some superficial things I wanted to ask about right now. Didn't finish reviewing every detail.

File: src/poseidon/poseidon_ref.rs (hunk 4 above)
Comment (BoyuanFeng) [reply src/poseidon/poseidon_ref.rs]: It is concatenated with hash input as the final hash input.

For example, if arity is 2, we have 2+1 hash inputs.

Comment (tsunrise) [reply src/poseidon/poseidon_ref.rs]: for now, domain tag is just `2^ARITY - 1`

File: src/poseidon/matrix.rs (hunk 3 above)
Comment (tsunrise) [reply src/poseidon/matrix.rs:94]: yes all MDS matrices should be square

File: src/poseidon/matrix.rs (hunk 2 above)
Comment (tsunrise) [reply src/poseidon/matrix.rs:8]: make sense (for better cache performance)

File: src/poseidon/matrix.rs (hunk 3 above)
Comment (bhgomes) [reply src/poseidon/matrix.rs:94]: Then, we should notate this and make sure that we are using this fact everywhere in the interface and the implementation.

File: src/poseidon/poseidon_ref.rs (hunk 4 above)
Comment (bhgomes) [reply src/poseidon/poseidon_ref.rs]: Is this in the original spec? I thought you were supposed to zero-pad the input like this:

```
fn hash(x: T, y: T) -> T {
    poseidon_permutation_of_width_3(0, x, y)
}
```

**Comment by tsunrise** [comment]
moved to https://github.com/Manta-Network/manta-rs/pull/57

**Pull request closed without merging**